pinot-core/src/main/java/org/apache/pinot/core/query/reduce/DistinctDataTableReducer.java

```diff
@@ -115,18 +140,31 @@ public void reduceAndSetResults(String tableName, DataSchema dataSchema, Map<Ser
       }
       resultSet.add(distinctRow);
     }
+    return new SelectionResults(Arrays.asList(columnNames), resultSet);
+  }
 
-    // Up until now, we have treated DISTINCT similar to another aggregation function even in terms
-    // of the result from function and merging results.
-    // However, the DISTINCT query is just another SELECTION style query from the user's point
-    // of view and will return one or records in the result table for the column(s) selected and so
-    // for that reason, response from broker should be a selection query result.
-    brokerResponseNative.setSelectionResults((new SelectionResults(Arrays.asList(columnNames), resultSet)));
+  private ResultTable reduceToResultTable(DistinctTable distinctTable) {
+    List<Object[]> resultSet = new ArrayList<>(distinctTable.size());
+    Iterator<Record> iterator = distinctTable.iterator();
+    while (iterator.hasNext()) {
+      Record record = iterator.next();
+      resultSet.add(record.getValues());
+    }
+    DataSchema finalDataSchema = getResultTableDataSchema();
+    return new ResultTable(finalDataSchema, resultSet);
   }
 
   private String[] getDistinctColumns() {
     return _brokerRequest.getAggregationsInfo().get(0).getAggregationParams()
         .get(FunctionCallAstNode.COLUMN_KEY_IN_AGGREGATION_INFO)
         .split(FunctionCallAstNode.DISTINCT_MULTI_COLUMN_SEPARATOR);
   }
+
+  private DataSchema getResultTableDataSchema() {
+    String[] columns = getDistinctColumns();
+    DataSchema.ColumnDataType[] columnDataTypes = new DataSchema.ColumnDataType[columns.length];
+    // TODO: there's no way currently to get the data types of the distinct columns
```

Comment:
You can do distinctTable.getDataSchema() and that should have both column names and types.

Reply:
Kept this method for the empty results case. Implemented your suggestion for the regular case